Remove metadata from snapshot/clone/parent RBD image

pkalever opened the pull request:
<!-- Please take a look at our [Contributing](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#Code-contribution-workflow)
documentation before submitting a Pull Request!
Thank you for contributing to ceph-csi! -->

# Describe what this PR does #

Remove metadata on images for various scenarios:
* Unset the PVC metadata on the rbd image created for the snapshot
* Unset the snapshot metadata from the rbd image created from the snapshot
* Unset the parent PVC metadata on the temp clone rbd image
* Also add various testcases to validate unset of metadata
## Related issues ##

Fixes: #2970

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

- `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)
- `/retest all`: run this in case the CentOS CI failed to start/report any test
  progress or results

</details>

Commit:
rbd: add unset volume/snapshot metadata utility functions

Added
GetVolumeMetadataKeys()
GetSnaoshotMetadataKeys()
unsetVolumeMetadata() and
unsetSnapshotMetadata()

functions.

Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>

Commit:
rbd: unset PVC metadata on CreateSnapshot

Unset the PVC metadata on the rbd image created for the snapshot

Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>

Commit:
rbd: unset snapshot metadata on CreateVolume From snapshot

Unset the snapshot metadata from the rbd image created from the snapshot

Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>

Commit:
rbd: unset parent PVC metadata on CreateVolume From Volume

Unset the parent PVC metadata on the temp clone rbd image

Fixes: #2970
Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -1934,6 +1934,24 @@ func (rv *rbdVolume) setVolumeMetadata(parameters map[string]string) error {
 	return nil
 }
 
```
this commit has _untility_ in the title :thinking:

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -1934,6 +1934,24 @@ func (rv *rbdVolume) setVolumeMetadata(parameters map[string]string) error {
 }
 
+// unsetVolumeMetadata unset PV/PVC/PVCNamespace metadata on RBD image.
+func (rv *rbdVolume) unsetVolumeMetadata() error {
```
in order to keep aligned with the parameters passed to `setVolumeMetadata()`, it would be better to pass a `[]string` with the keys instead of fetching them inside the function

pkalever replied:
That would be an unnecessary duplication of  `k8s.GetVolumeMetadataKeys()` everywhere. I would rather prefer it to be in one place.

nixpanic replied:
it is out of balance with `setVolumeMetadata()`, so I prefer that you pass the keys. Possibly other metadata gets added later on, which does not get retrieved from the k8s API.

pkalever replied:
I don't know how it becomes out of balance with `setVolumeMetadata()`, both the function are getting the keys with in the function. And in the future if the other metadata gets added later on it applies to `setVolumeMetadata()` too. So I would still let it just be it for now.

nixpanic replied:
Indeed! additional metadata should be passed as an argument to `setVolumeMetadata()`, so the caller of `unsetVolumeMetadata()` should have that as well. The `rbdVolume` functions are an API for external callers, and there should not be a dependency on hard-coded metadata keys that are not internal to the `rbdVolume`.

pkalever replied:
@nixpanic Done now, PTAL

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -1946,3 +1964,22 @@ func (rv *rbdVolume) setSnapshotMetadata(parameters map[string]string) error {
+// unsetSnapshotMetadata unset snapshot-name/snapshot-namespace/snapshotcontent-name metadata
+// on RBD image.
+func (rv *rbdVolume) unsetSnapshotMetadata() error {
+	for _, key := range k8s.GetSnaoshotMetadataKeys() {
```
`GetSnaoshotMetadataKeys`, _Snaoshot_?

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -1946,3 +1964,22 @@ func (rv *rbdVolume) setSnapshotMetadata(parameters map[string]string) error {
 	return nil
 }
+
+// unsetSnapshotMetadata unset snapshot-name/snapshot-namespace/snapshotcontent-name metadata
```
you probably do not need this function if you pass the keys as a `[]string`. It is the same as `unsetVolumeMetadata()`?

pkalever replied:
Yeah, let's have unsetSnapshotMetadata like unsetVolumeMetadata separately, this will avoid certain confusion too.

nixpanic replied:
the function is almost completely duplicated, if there really is a need for separate functions (hint: there probably is not), use a helper to improve maintainability.

pkalever replied:
In addition to my previous comment,
Calling unsetVolumeMetadata() for Snapshot metadata looks a bit awkward to me, CodeReadabilty++

nixpanic replied:
just rename the function to `SetMetadataList(data map[string]string)` and `UnsetMetadataList(keys []string)`? The object for both Snapshot and Volume are `rbdVolume`, so no need to have Volume or Snapshot in the function names. CodeReadabilty++ and CodeMaintainability++

pkalever replied:
@nixpanic Done now, PTAL

nixpanic commented on internal/util/k8s/parameters.go:
```diff
@@ -100,3 +109,12 @@ func GetSnapshotMetadata(parameters map[string]string) map[string]string {
 }
+
+// GetSnapshotMetadataKeys return snapshot metadata keys.
+func GetSnaoshotMetadataKeys() []string {
```
_Snaoshot_

nixpanic commented on e2e/rbd.go:
```diff
@@ -675,6 +676,36 @@ var _ = Describe("RBD", func() {
+					fmt.Sprintf("rbd image-meta get %s --image=%s %s",
+						rbdOptions(defaultRBDPool), imageList[0], pvcNameKey),
+					rookNamespace)
+				if err == nil || !strings.Contains(err.Error(), "command terminated with exit code 2") ||
```
this is a very long if-check, maybe use a helper function as it is repeated often?

pkalever reviewed (commented):
Thanks for the review @nixpanic :-)

Madhu-1 commented on e2e/rbd.go:
```diff
@@ -675,6 +685,33 @@ var _ = Describe("RBD", func() {
 					e2elog.Failf("expected volSnapContentName %q got %q", content.Name, volSnapContentName)
 				}
 
+				// make sure we had unset the PVC metadata on the rbd image created
+				// for the snapshot
+				pvcName, stdErr, err := execCommandInToolBoxPod(f,
+					fmt.Sprintf("rbd image-meta get %s --image=%s %s",
+						rbdOptions(defaultRBDPool), imageList[0], pvcNameKey),
+					rookNamespace)
+				if isKeyExist(err, stdErr) {
+					e2elog.Failf("PVC name found on %s/%s %s=%s: err=%v stdErr=%q",
+						rbdOptions(defaultRBDPool), imageList[0], pvcNameKey, pvcName, err, stdErr)
+				}
+				pvcNamespace, stdErr, err := execCommandInToolBoxPod(f,
+					fmt.Sprintf("rbd image-meta get %s --image=%s %s",
+						rbdOptions(defaultRBDPool), imageList[0], pvcNamespaceKey),
+					rookNamespace)
+				if isKeyExist(err, stdErr) {
+					e2elog.Failf("PVC namespace found on %s/%s %s=%s: err=%v stdErr=%q",
+						rbdOptions(defaultRBDPool), imageList[0], pvcNamespaceKey, pvcNamespace, err, stdErr)
+				}
+				pvName, stdErr, err := execCommandInToolBoxPod(f,
+					fmt.Sprintf("rbd image-meta get %s --image=%s %s",
+						rbdOptions(defaultRBDPool), imageList[0], pvNameKey),
+					rookNamespace)
+				if isKeyExist(err, stdErr) {
+					e2elog.Failf("PV name found on %s/%s %s=%s: err=%v stdErr=%q",
+						rbdOptions(defaultRBDPool), imageList[0], pvNameKey, pvName, err, stdErr)
+				}
+
```
Do list of metadata with `rbd image-meta ls` and check anything related to PVC is present.

pkalever replied:
List will dump all the metadata at once
```
sh-4.4$ rbd image-meta list replicapool/csi-vol-29bf8a0a-ba45-11ec-bd09-fa9af733450f
There are 3 metadata on this image:

Key                               Value                                   
csi.storage.k8s.io/pv/name        pvc-890cb9a5-a5f8-42d9-825b-de780377c6ab
csi.storage.k8s.io/pvc/name       rbd-pvc                                 
csi.storage.k8s.io/pvc/namespace  default                                 

```

String.contains("csi.storage.k8s.io/pvc/name", "csi.storage.k8s.io/pvc/namespace  default")
Might return success. All I'm saying is, let's not take that chance.

@Madhu-1 WDYT?

Madhu-1 replied:
IMO Instead of getting separately multiple times, the listing is easy, list once and check it.

Madhu-1 commented on e2e/rbd.go:
```diff
@@ -686,6 +723,251 @@ var _ = Describe("RBD", func() {
+					e2elog.Failf("found snapshotcontent name %s/%s %s=%s: err=%v stdErr=%q",
+						rbdOptions(defaultRBDPool), imageList[0], volSnapContentNameKey,
+						volSnapContentName, err, stdErr)
+				}
```
use `ls` to list the metadata and check anything related to snapshot exists.

Madhu-1 commented on e2e/rbd.go:
```diff
@@ -686,6 +723,251 @@ var _ = Describe("RBD", func() {
 			})
 
+			By("verify that we unset snapshot metadata on CreateVolume from snapshot", func() {
+				const (
```
instead of adding a  new test please extend the existing test for the below 2 reasons
* As adding new tests for this one increases the test timings in the CI.
* This needs to be verified along with the CreateVolumeSnapshot test we don't need to have a separate test to validate the snapshot metadata.

Madhu-1 commented on e2e/rbd.go:
```diff
@@ -686,6 +723,251 @@ var _ = Describe("RBD", func() {
+				}
+			})
+
+			By("verify that we unset parent PVC metadata on temp clone image", func() {
```
instead of adding a new test please extend the existing test for the below 2 reasons

* As adding new tests for this one increases the test timings in the CI.
* This needs to be verified along with the CreateVolume test we don't need to have a separate test to validate the older metadata.

Madhu-1 commented on internal/rbd/clone.go:
```diff
@@ -190,6 +190,12 @@ func (rv *rbdVolume) doSnapClone(ctx context.Context, parentVol *rbdVolume) erro
 	}
+	err = tempClone.unsetVolumeMetadata()
+	if err != nil {
+		log.ErrorLog(ctx, "failed to unset volume metadata on temp clone: %v", err)
```
log image name.

Madhu-1 commented on internal/rbd/clone.go:
```diff
@@ -190,6 +190,12 @@ func (rv *rbdVolume) doSnapClone(ctx context.Context, parentVol *rbdVolume) erro
 	if err != nil {
 		return err
 	}
+	err = tempClone.unsetVolumeMetadata()
```
the tempClone should contain the metadata same as the image it's created for.

pkalever replied:
As discussed, let's do this on a need basis.

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -588,6 +588,12 @@ func (cs *ControllerServer) createVolumeFromSnapshot(
 	}
+	err = rbdVol.unsetSnapshotMetadata()
+	if err != nil {
+		log.ErrorLog(ctx, "failed to unset snapshot metadata on rbd image: %v", err)
```
log rbdVol details

Madhu-1 commented on internal/rbd/rbd_util.go:
```diff
@@ -1934,6 +1934,24 @@ func (rv *rbdVolume) setVolumeMetadata(parameters map[string]string) error {
+// unsetVolumeMetadata unset PV/PVC/PVCNamespace metadata on RBD image.
+func (rv *rbdVolume) unsetVolumeMetadata() error {
+	for _, key := range k8s.GetVolumeMetadataKeys() {
+		value, err := rv.GetMetadata(key)
+		if errors.Is(err, librbd.ErrNotFound) {
+			continue
+		} else if err != nil {
+			return err
+		}
```
instead of getting the keys one by one, do `list` and Compare and remove it.

Madhu-1 commented on internal/rbd/rbd_util.go:
```diff
@@ -1934,6 +1934,24 @@ func (rv *rbdVolume) setVolumeMetadata(parameters map[string]string) error {
+		}
+		err = rv.RemoveMetadata(key)
+		if err != nil {
+			return fmt.Errorf("failed to unset metadata key %q, value %q on image: %w", key, value, err)
```
log rv vol details

Madhu-1 commented on internal/rbd/rbd_util.go:
```diff
@@ -1946,3 +1964,22 @@ func (rv *rbdVolume) setSnapshotMetadata(parameters map[string]string) error {
+// on RBD image.
+func (rv *rbdVolume) unsetSnapshotMetadata() error {
+	for _, key := range k8s.GetSnapshotMetadataKeys() {
+		value, err := rv.GetMetadata(key)
+		if errors.Is(err, librbd.ErrNotFound) {
+			continue
+		} else if err != nil {
```
instead of getting the keys one by one, do list and Compare and remove it.

Madhu-1 commented on internal/rbd/rbd_util.go:
```diff
@@ -1946,3 +1964,22 @@ func (rv *rbdVolume) setSnapshotMetadata(parameters map[string]string) error {
+		}
+		err = rv.RemoveMetadata(key)
+		if err != nil {
+			return fmt.Errorf("failed to unset metadata key %q, value %q on image: %w", key, value, err)
```
log image details

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

pkalever commented:
/retest ci/centos/k8s-e2e-external-storage/1.22

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -1934,6 +1946,29 @@ func (rv *rbdVolume) setVolumeMetadata(parameters map[string]string) error {
+
+	for k, v := range list {
+		for _, key := range k8s.GetVolumeMetadataKeys() {
+			if strings.Compare(k, key) != 0 {
```
why `strings.Compare` and not just `k == key`?

pkalever replied:
I was over thinking may be :-)

pkalever commented:
/retest all

pkalever commented:
@Mergifyio rebase

pkalever commented:
@Mergifyio rebase

pkalever commented:
@Mergifyio rebase

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

pkalever commented:
/retest ci/centos/k8s-e2e-external-storage/1.21

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

Commit:
cleanup: simplify setVolumeMetadata and rename it

Move k8s.GetVolumeMetadata() out of setVolumeMetadata() and rename it to
setAllMetadata() so that the same can be used for setting volume and
snapshot metadata.

Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>

pkalever commented:
@Madhu-1 @nixpanic @ceph/ceph-csi-contributors please review this changes. Thanks!

pkalever commented:
/retest all

pkalever commented:
/retest ci/centos/k8s-e2e-external-storage/1.22

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.21

yati1998 reviewed (commented):
LGTM, reserving merge rights for others.

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -1932,3 +1944,26 @@ func (rv *rbdVolume) setAllMetadata(parameters map[string]string) error {
+			err = rv.RemoveMetadata(key)
+			if err != nil && !strings.Contains(err.Error(), "No such file or directory") {
+				return fmt.Errorf("failed to unset metadata key %q, value %q on image %s/%s: %w",
+					k, v, rv.Pool, rv.RbdImageName, err)
```
just pass `rv` in `%q` as that will return a string with pool/namespace/imagename

nixpanic commented on e2e/rbd.go:
```diff
@@ -191,6 +191,15 @@ func validateRBDImageCount(f *framework.Framework, count int, pool string) {
 	}
 }
 
+func isKeyExist(err error, stdErr string) bool {
```
I do not get what this function does... the name does not describe the contents?

nixpanic commented on e2e/rbd.go:
```diff
@@ -652,6 +661,33 @@ var _ = Describe("RBD", func() {
+				// make sure we had unset the PVC metadata on the rbd image created
+				// for the snapshot
+				pvcName, stdErr, err := execCommandInToolBoxPod(f,
+					fmt.Sprintf("rbd image-meta get %s --image=%s %s",
```
please use a helper function for `rbd image-meta get ....`

nixpanic reviewed (changes requested):
just a few nits, looks good otherwise

Commit:
e2e: add various testcases to validate unset of metadata

Validate that we:
* Unset the PVC metadata on the rbd image created for the snapshot
* snapshot metadata on CreateVolume from snapshot

Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>

pkalever commented:
@nixpanic addressed your review comments now, PTAL. Thanks!

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

pkalever commented:
@Mergifyio rebase

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

Madhu-1 commented on internal/rbd/clone.go:
```diff
@@ -190,6 +191,13 @@ func (rv *rbdVolume) doSnapClone(ctx context.Context, parentVol *rbdVolume) erro
 	}
+	err = tempClone.unsetAllMetadata(k8s.GetSnapshotMetadataKeys())
+	if err != nil {
+		log.ErrorLog(ctx, "failed to unset volume metadata on temp clone image %s/%s: %v",
```
```suggestion
log.ErrorLog(ctx, "failed to unset volume metadata on temp clone image %s %v",
			tempClone, err)
```
tempClone already implements String interface,

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -588,6 +590,13 @@ func (cs *ControllerServer) createVolumeFromSnapshot(
 	}
+	err = rbdVol.unsetAllMetadata(k8s.GetSnapshotMetadataKeys())
+	if err != nil {
+		log.ErrorLog(ctx, "failed to unset snapshot metadata on rbd image %s/%s: %v",
```
same here as well.

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -963,7 +972,7 @@ func (cs *ControllerServer) ValidateVolumeCapabilities(
 
 // CreateSnapshot creates the snapshot in backend and stores metadata in store.
-// nolint:cyclop // TODO: reduce complexity
+// nolint:gocyclo,cyclop // TODO: reduce complexity.
```
adding nolint is not helpful we will forget about it once PR is merged and no use of having linter checks in CI isn't it?

pkalever replied:
This function is already complex and at its best-simplified level. You can see it already has `cyclop`.
Excuses apart, we should consider scrubbing such complex functions sometimes.

Madhu-1 commented on internal/rbd/rbd_util.go:
```diff
@@ -1934,13 +1945,22 @@ func (rv *rbdVolume) setVolumeMetadata(parameters map[string]string) error {
-			return fmt.Errorf("failed to set metadata key %q, value %q on image: %w", k, v, err)
+// unsetAllMetadata unset all the metadata from arg keys on RBD image.
+func (rv *rbdVolume) unsetAllMetadata(keys []string) error {
+	list, err := rv.ListMetadata()
+	if err != nil {
+		return fmt.Errorf("failed to list metadata image %s/%s: %w", rv.Pool, rv.RbdImageName, err)
+	}
+
+	for k, v := range list {
+		for _, key := range keys {
+			if k != key {
+				continue
+			}
+			err = rv.RemoveMetadata(key)
+			if err != nil && !strings.Contains(err.Error(), "No such file or directory") {
+				return fmt.Errorf("failed to unset metadata key %q, value %q on %q: %w", k, v, rv, err)
+			}
```
To unset why we need to list and remove it? just do RemoveMetadata and ignore not/no found error?

pkalever commented:
@Mergify rebase

pkalever commented:
@Madhu-1 @nixpanic PTAL

Madhu-1 commented on internal/rbd/rbd_util.go:
```diff
@@ -1934,13 +1945,12 @@ func (rv *rbdVolume) setVolumeMetadata(parameters map[string]string) error {
+func (rv *rbdVolume) unsetAllMetadata(keys []string) error {
+	for _, key := range keys {
+		err := rv.RemoveMetadata(key)
+		if err != nil && !strings.Contains(err.Error(), "No such file or directory") {
+			return fmt.Errorf("failed to unset metadata key %q on %q: %w", key, rv, err)
```
```suggestion
if err != nil && !errors.Is(err, rados.ErrNotFound) {
return fmt.Errorf("failed to unset metadata key %q on %q: %w", key, rv, err)
}
```

can we do like this?

pkalever replied:
@Madhu-1 This is what I'm talking about in the call, can I revert this?
rbd upgrade test is failing with:
```
May 10 13:01:46.985: INFO: At 2022-05-10 12:51:48 +0100 BST - event for rbd-pvc-restore: {rbd.csi.ceph.com_csi-rbdplugin-provisioner-57c4f55c9c-thjpc_96186737-0ddd-4b5f-bef5-2df83e7ac0d1 } ProvisioningFailed: failed to provision volume with StorageClass "csi-rbd-sc": rpc error: code = Unknown desc = failed to unset metadata key "csi.storage.k8s.io/volumesnapshot/name" on "replicapool/csi-vol-92685dd1-d057-11ec-83c4-862ba5925cfb": rbd: ret=-2, No such file or directory
```

Madhu-1 reviewed (commented):
small suggestion on error handling

pkalever commented:
@Madhu-1 @nixpanic PTAL

Madhu-1 commented on internal/rbd/clone.go:
```diff
@@ -188,6 +189,12 @@ func (rv *rbdVolume) doSnapClone(ctx context.Context, parentVol *rbdVolume) erro
 	if err != nil {
 		return err
 	}
+	err = tempClone.unsetAllMetadata(k8s.GetSnapshotMetadataKeys())
```
you should be resetting the parent volume metadata here isn't it? or its snapshot metadata as its done here

pkalever replied:
YEs, good catch. Looks like it got messed up over various revisions of the patch sets. Thanks!

Madhu-1 commented on internal/rbd/rbd_util.go:
```diff
@@ -1675,6 +1674,18 @@ func (ri *rbdImage) RemoveMetadata(key string) error {
 
+// ListMetadata returns a map containing all metadata assigned to the RBD
+// image.
+func (ri *rbdImage) ListMetadata() (map[string]string, error) {
```
is it used anywhere? if not please consider removing it.

pkalever replied:
Sure! I just kept it reserving for future :-)

Madhu-1 commented on internal/rbd/clone.go:
```diff
@@ -188,6 +189,12 @@ func (rv *rbdVolume) doSnapClone(ctx context.Context, parentVol *rbdVolume) erro
 	if err != nil {
 		return err
 	}
+	err = tempClone.unsetAllMetadata(k8s.GetVolumeMetadataKeys())
```
we should unset metadata at line 215 as if the `unsetAllMetadata` fails the tempClone need to be removed.

pkalever replied:
ok

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -589,6 +591,12 @@ func (cs *ControllerServer) createVolumeFromSnapshot(
 
 		return err
 	}
+	err = rbdVol.unsetAllMetadata(k8s.GetSnapshotMetadataKeys())
```
not related to this PR I see a cleanup bug here. if the unsetAllMetadata or copyEncryptionConfig or expand fails we need to remove the clone. can you please open an issue for this one?

pkalever replied:
Sure Madhu

pkalever replied:
https://github.com/ceph/ceph-csi/issues/3103

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -1072,9 +1080,14 @@ func (cs *ControllerServer) CreateSnapshot(
 	// Update the metadata on snapshot not on the original image
 	rbdVol.RbdImageName = rbdSnap.RbdSnapName
 
+	err = rbdVol.unsetAllMetadata(k8s.GetVolumeMetadataKeys())
```
same here as well.

pkalever replied:
https://github.com/ceph/ceph-csi/issues/3103

pkalever commented:
/retest ci/centos/upgrade-tests-rbd

yati1998 commented on e2e/rbd.go:
```diff
@@ -191,6 +191,21 @@ func validateRBDImageCount(f *framework.Framework, count int, pool string) {
+	return fmt.Sprintf("rbd image-meta get %s --image=%s %s", rbdOptions(pool), image, key)
+}
+
+// checkKeyError check for error conditions returned by get image-meta key,
```
checkKeyError -> checkGetKeyError

pkalever replied:
Done now, thanks!

yati1998 reviewed (dismissed):
Just a small nit, else looks good to me.

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

pkalever commented:
@Mergifyio rebase

pkalever commented:
/retest ci/centos/upgrade-tests-rbd

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

pkalever commented:
/retest ci/centos/upgrade-tests-rbd

pkalever commented:
/retest ci/centos/upgrade-tests-rbd

pkalever commented:
/retest ci/centos/upgrade-tests-rbd

pkalever commented:
/retest ci/centos/upgrade-tests-rbd

pkalever commented:
/retest ci/centos/upgrade-tests-rbd

pkalever commented:
/retest ci/centos/upgrade-tests-rbd

pkalever commented:
/retest ci/centos/upgrade-tests-rbd

pkalever commented:
/retest ci/centos/upgrade-tests-rbd

pkalever commented:
/retest ci/centos/upgrade-tests-rbd

pkalever commented:
/retest ci/centos/upgrade-tests-rbd

pkalever commented:
/retest ci/centos/upgrade-tests-rbd

nixpanic commented:
This can be rebased with Mergify, just have CI finish testing and merging other PRs.

nixpanic commented:
@mergifyio rebase

pkalever commented:
Diff:
V3: https://github.com/ceph/ceph-csi/pull/3005#discussion_r866707138 is not working, so replaced errno with string comparison for now. Added a TODO, will address in future PRs

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21
/retest ci/centos/mini-e2e-helm/k8s-1.23
```
May 11 10:33:45.243: FAIL: backend images not matching kubernetes resource count,image count 23 kubernetes resource count 21
```

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

nixpanic commented:
/retest all

nixpanic commented:
> /retest all

It seems the CI jobs got stuck somehow. This PR is the 1st in the queue to be merged, hopefully things pass quickly.

nixpanic commented:
@mergifyio unqueue

#3070 has passed CI, and should be able to get merged immediately. Let's try the new `dequeue` command

nixpanic commented:
@mergifyio rebase

nixpanic commented:
@mergifyio requeue

The other PR got merged, so this one is next now. Restarted CI jobs after rebasing.

nixpanic commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

nixpanic commented:
> /retest ci/centos/mini-e2e-helm/k8s-1.23

Failed with:
```
    May 12 09:46:29.815: backend images not matching kubernetes resource count,image count 23 kubernetes resource count 21
```

nixpanic commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

nixpanic commented:
> /retest ci/centos/mini-e2e-helm/k8s-1.23

Failed with:
```
May 12 10:50:53.103: FAIL: backend images not matching kubernetes resource count,image count 12 kubernetes resource count 11
```

nixpanic commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

nixpanic commented:
> /retest ci/centos/mini-e2e-helm/k8s-1.23

Failed with:
```
May 12 12:14:55.114: FAIL: backend images not matching kubernetes resource count,image count 22 kubernetes resource count 21
```
